File: Cpp/622_Design_Circular_Queue.cpp
class MyCircularQueue {
public:
    /** Initialize your data structure here. Set the size of the queue to be k. */
    MyCircularQueue(int k) : q(k+1), q_size(k+1), front(0), rear(0) {
        
    }
    
    /** Insert an element into the circular queue. Return true if the operation is successful. */
    bool enQueue(int value) {
        if(isFull()) return false;

        q[rear] = value;
        rear = (rear + 1) % (q_size);
        return true;
    }
    
    /** Delete an element from the circular queue. Return true if the operation is successful. */
    bool deQueue() {
        if(isEmpty()) return false;
        front = (front + 1) % q_size;
        return true;
    }
    
    /** Get the front item from the queue. */
    int Front() {
        if(isEmpty()) return -1;
        return q[front];
    }
    
    /** Get the last item from the queue. */
    int Rear() {
        if(isEmpty()) return -1;
        return q[(rear - 1 + q_size) % q_size];
    }
    
    /** Checks whether the circular queue is empty or not. */
    bool isEmpty() {
        return front == rear;
    }
    
    /** Checks whether the circular queue is full or not. */
    bool isFull() {
        return (rear + 1) % q_size == front;
    }
private:
    vector<int> q;
    int q_size;
    int front;
    int rear;
};

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * MyCircularQueue* obj = new MyCircularQueue(k);
 * bool param_1 = obj->enQueue(value);
 * bool param_2 = obj->deQueue();
 * int param_3 = obj->Front();
 * int param_4 = obj->Rear();
 * bool param_5 = obj->isEmpty();
 * bool param_6 = obj->isFull();
 */

File: Kotlin/622_Design_Circular_Queue.kt
class MyCircularQueue(k: Int) {
    private val q = IntArray(k + 1) { -1 }
    private var front = 0
    private var rear = 0
    
    fun enQueue(value: Int): Boolean {
        if(isFull()) return false

        q[rear] = value
        rear = (rear + 1) % (q.size)
        
        return true
    }

    fun deQueue(): Boolean {
        if(isEmpty()) return false
        
        front = (front + 1) % q.size
        
        return true
    }

    fun Front(): Int {
        if(isEmpty()) return -1
        
        return q[front]
    }

    fun Rear(): Int {
        if(isEmpty()) return -1
        
        return q[(rear - 1 + q.size) % q.size]
    }

    fun isEmpty(): Boolean {
        return front == rear
    }

    fun isFull(): Boolean {
        return (rear + 1) % q.size == front
    }

}

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * var obj = MyCircularQueue(k)
 * var param_1 = obj.enQueue(value)
 * var param_2 = obj.deQueue()
 * var param_3 = obj.Front()
 * var param_4 = obj.Rear()
 * var param_5 = obj.isEmpty()
 * var param_6 = obj.isFull()
 */